### ISSUE TITLE
Adapter payloads arch and size problems

### ISSUE THREAD

**Issue by dwelch-r7** [rapid7/metasploit-framework#16582]
Problem discovered while running `./tools/modules/update_payload_cached_sizes.rb` but can be replicated manually by `use`ing the payloads in question and `generate`ing them

CC @smcintyre-r7 

Stacktrace of the result being too large after being adapted
```
/Users/user/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-powershell-0.1.96/lib/rex/powershell/command.rb:408:in `cmd_psh_payload': Powershell command length is greater than the command line maximum (8192 characters) (Rex::Powershell::Exceptions::PowershellCommandLengthError)
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/exploit/powershell.rb:250:in `rescue in cmd_psh_payload'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/exploit/powershell.rb:243:in `cmd_psh_payload'
        from /Users/user/Documents/code/metasploit-framework/modules/payloads/adapters/cmd/windows/powershell.rb:30:in `generate'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/payload.rb:299:in `generate_complete'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/encoded_payload.rb:118:in `generate_raw'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/encoded_payload.rb:74:in `generate'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/encoded_payload.rb:24:in `create'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/base/simple/payload.rb:52:in `generate_simple'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/base/simple/payload.rb:139:in `generate_simple'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/util/payload_cached_size.rb:118:in `block in is_dynamic?'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/util/payload_cached_size.rb:117:in `map'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/util/payload_cached_size.rb:117:in `is_dynamic?'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/util/payload_cached_size.rb:127:in `is_cached_size_accurate?'
        from ./tools/modules/update_payload_cached_sizes.rb:33:in `block in <main>'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/module_set.rb:272:in `block in each_module_list'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/module_set.rb:245:in `each'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/module_set.rb:245:in `each_module_list'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/module_set.rb:92:in `each_module'
        from ./tools/modules/update_payload_cached_sizes.rb:29:in `<main>'
/Users/user/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-powershell-0.1.96/lib/rex/powershell/command.rb:408:in `cmd_psh_payload': Powershell command length is greater than the command line maximum (8192 characters) (Rex::Powershell::Exceptions::PowershellCommandLengthError)
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/exploit/powershell.rb:244:in `cmd_psh_payload'
        from /Users/user/Documents/code/metasploit-framework/modules/payloads/adapters/cmd/windows/powershell.rb:30:in `generate'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/payload.rb:299:in `generate_complete'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/encoded_payload.rb:118:in `generate_raw'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/encoded_payload.rb:74:in `generate'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/encoded_payload.rb:24:in `create'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/base/simple/payload.rb:52:in `generate_simple'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/base/simple/payload.rb:139:in `generate_simple'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/util/payload_cached_size.rb:118:in `block in is_dynamic?'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/util/payload_cached_size.rb:117:in `map'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/util/payload_cached_size.rb:117:in `is_dynamic?'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/util/payload_cached_size.rb:127:in `is_cached_size_accurate?'
        from ./tools/modules/update_payload_cached_sizes.rb:33:in `block in <main>'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/module_set.rb:272:in `block in each_module_list'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/module_set.rb:245:in `each'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/module_set.rb:245:in `each_module_list'
        from /Users/user/Documents/code/metasploit-framework/lib/msf/core/module_set.rb:92:in `each_module'
        from ./tools/modules/update_payload_cached_sizes.rb:29:in `<main>'
```

**Comment by smcintyre-r7** [comment]
The first looks pretty easy to fix. The adapter should have a size check [like the Python one does](https://github.com/rapid7/metasploit-framework/blob/e386dad7e9017cbbfc371bf999d8a034bd5827a8/modules/payloads/adapters/cmd/unix/python.rb#L26) to ensure that it's not used with the stageless payloads since those would be too large.

Can you help me reproduce the second one though? The architecture *should be* ARCH_CMD, once it's been adapted, so I'm guessing somewhere there was an assumption made that it would never be that value which needs to be updated.

Also if you want to file a separate ticket for that second one, I can submit the patch for the first now. Not sure how long the second one will take to fix.

**Comment by smcintyre-r7** [comment]
Fixed the first issue over here in my [fix/issue/16582](https://github.com/zeroSteiner/metasploit-framework/tree/fix/issue/16582) branch. I landed on 3,000 because the space is largely dependent on the datastore options like if the bypasses are prepended, the method, inner / outer encoding settings etc. 4,000 worked reliably but IIRC there's not alot of payload in the size range of 3,000-4,000 so it seemed better to err on the side of smaller. It's possible that the theoretical size adjustment ratio could also change due to updates to the rex-powershell repo.

I'll hold off on submitting until I can reproduce and fix the second.
